Add content_unformatted config to avoid formatting content of an element (fix #906)

arai-a opened the pull request:
This fixes #906

The issue is that `unformatted` config is applied to following 3 places:
  * newline before the element
  * content of the element
  * newline after the element

while it's useful for inline elements, when one want to disable formatting only the content of the content, keeping newlines before/after it, the config cannot be used.

I added new `content_unformatted` config, that does only "content of the element" part.
if `content_unformatted` one add "style" and "script" to this, their content is not formatted, while they still have newlines before/after them.

config file:

```
{
    "content_unformatted": [ "script", "style" ]
}
```

output

```
$ node ./js/bin/html-beautify.js   --config ~/Desktop/conf ~/Desktop/b.html
<body>
    <h1>Heading</h1>
    <script>var formatMe = function() { return false; };</script>
    <style>.format-disabled { display: none; } </style>
</body>
```


I also added commandline option `--content_unformatted`, or `-T` (the first character in 'content_unformatted' that wasn't used...), that can be used like `--unformatted`.

```
$ node ./js/bin/html-beautify.js -T script ~/Desktop/b.html
<body>
    <h1>Heading</h1>
    <script>var formatMe = function() { return false; };</script>
    <style>
        .format-disabled {
            display: none;
        }
    </style>
</body>

$ node ./js/bin/html-beautify.js -T style ~/Desktop/b.html
<body>
    <h1>Heading</h1>
    <script>
        var formatMe = function() {
            return false;
        };
    </script>
    <style>.format-disabled { display: none; } </style>
</body>

$ node ./js/bin/html-beautify.js -T script -T style ~/Desktop/b.html
<body>
    <h1>Heading</h1>
    <script>var formatMe = function() { return false; };</script>
    <style>.format-disabled { display: none; } </style>
</body>
```

bitwiseman commented:
@arai-a - Thanks for submitting. 
Please rebase the latest changes. 
Also, look at the tests for #1081 and make sure you handle those inputs, as well. 
Your change looks simpler than that one as well.

Finally, please format your multi-line test inputs or outputs like this to make them easier to understand:

```js
        tests: [{
            fragment: true,
            input: '<html><body><h1>A</h1><script>if(1){f();}</script><style>.a{display:none;}</style></body></html>',
            output: [
                '<html>',
                '<body>',
                '    <h1>A</h1>',
                '    <script>if(1){f();}</script>',
                '    <style>.a{display:none;}</style>',
                '</body>',
                '',
                '</html>'
            ]
        }]
```

arai-a commented:
rebased onto latest master and added tests from #1081 and it passed without any changes.
thank you!

bitwiseman commented on js/lib/beautify-html.js:
```diff
@@ -161,6 +163,7 @@
             'acronym', 'address', 'big', 'dt', 'ins', 'strike', 'tt',
             'pre',
         ];
+        content_unformatted = options.content_unformatted || [];
```
Please move `'pre'` from `unformatted`, into the default for content_unformatted and add some tests for it based on #928.

bitwiseman commented on test/data/html/tests.js:
```diff
@@ -900,6 +900,52 @@ exports.test_data = {
+        name: "content_unformatted to prevent formatting content",
+        description: "",
+        options: [
+            { name: 'content_unformatted', value: "['script', 'style', 'p', 'span', 'br']" }
```
Add some examples with `<pre>` tag (as mentioned above).
Then please do the same inputs with the default  `content_unformatted`.  It is important to understand how this changes the outputs.

arai-a commented:
okay, moved pre to content_unformatted and added tests for pre and default options,
and also updated one pre-existing test to follow the change to pre

bitwiseman commented:
Great, thanks.  
There's been a question from #1081.   I need to establish whether this issue raised there is valid or not before we merge.  Other than that this looks good.

arai-a commented:
I don't see the missing newline issue, on osx, node v6.9.1.
I get expected result with cli.

arai-a commented:
tested also on debian, nodejs v4.6.1,
I get expected result.

bitwiseman commented:
Thank you for your contributio. I hope you will continue to contribute beyond this one PR.

arai-a commented:
thank you too for reviewing :)
It's my pleasure to be able to contribute to software I use daily.
